peer/peerdiscovery.go

```diff
@@ -0,0 +1,174 @@
+package peer
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/renproject/aw/transport"
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+
+	"go.uber.org/zap"
+)
+
+type SignatoryAndAddress struct {
+	Signatory id.Signatory
+	Address   wire.Address
+}
+
+func (sigAndAddr SignatoryAndAddress) Marshal(buf []byte, rem int) ([]byte, int, error) {
+	buf, rem, err := sigAndAddr.Signatory.Marshal(buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("marshal signatory: %v", err)
+	}
+	buf, rem, err = sigAndAddr.Address.Marshal(buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("marshal address: %v", err)
+	}
+	return buf, rem, err
+}
+func (sigAndAddr *SignatoryAndAddress) Unmarshal(buf []byte, rem int) ([]byte, int, error) {
+	buf, rem, err := (&sigAndAddr.Signatory).Unmarshal(buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("unmarshal signatory: %v", err)
+	}
+	buf, rem, err = sigAndAddr.Address.Unmarshal(buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("unmarshal address: %v", err)
+	}
+	return buf, rem, err
+}
+
+type DiscoveryClient struct {
+	opts DiscoveryOptions
+
+	transport *transport.Transport
+}
+
+func NewDiscoveryClient(opts DiscoveryOptions, transport *transport.Transport) *DiscoveryClient {
+	return &DiscoveryClient{
+		opts: opts,
+
+		transport: transport,
+	}
+}
+
+func (dc *DiscoveryClient) PeerDiscovery(ctx context.Context) {
+	var pingData [2]byte
+	binary.LittleEndian.PutUint16(pingData[:], dc.transport.Port())
+
+	msg := wire.Msg{
+		Version: wire.MsgVersion1,
+		Type:    wire.MsgTypePing,
+		Data:    pingData[:],
+	}
+
+	ticker := time.NewTicker(dc.opts.PingTimePeriod)
```

Comment:
Tickers must be stopped using `defer ticker.Stop()` otherwise resources are leaked.